src/mast/mod.rs

```diff
@@ -951,20 +994,35 @@ impl<B: Backend> Mast<B> {
                         _ => panic!("expected array type"),
                     };
 
+                    // bind generic value
+                    generics.bind(gen_name.to_string(), *size, sig_arg.span)?;
+
                     // store the inferred value for generic parameter
                     let gen_mty = MTypeInfo::new(&TyKind::BigInt, *span, Some(*size));
                     typed_fn_env.store_type(gen_name, &gen_mty)?;
 
                     // store concrete array type for the argument name
                     let arr_mty = MTypeInfo::new(arr_type, *span, None);
-                    typed_fn_env.store_type(&sig_arg.name.value, &arr_mty)?;
+                    typed_fn_env.store_type(arg_name, &arr_mty)?;
                 }
                 _ => {
                     let typ = type_info.typ.as_ref().expect("expected a value");
                     let cst = type_info.constant;
+
+                    if is_generic_parameter(arg_name) {
```

Comment:
maybe add a comment here:

> we expect `const TypeParam: Field` or `const normal_const: Field` or `normal_arg: SomeType`

although shouldn't that check be like in the parser? here we should just have a branch for that:

```rust
_ if is_generic_parameter(arg_name) && type_info.constant.is_some() => { /* ... */ }
_ => { /* ... */ }
```

Reply:
The check on the constant value is to make sure there is a constant value propagated.
Maybe this is unnecessary, as this should be checked at TAST.

Reply:
I think the parser can do that check no?

Reply:
yeah that seems unrelated to the MAST